# -*- coding: utf-8 -*-

## configuration for tox <http://tox.testrun.org/>

## tox automates running certain tasks within virtualenvs.  The following
## tox configuration outlines a basic setup for running unit tests and
## building sphinx docs in separate virtual environments.  Give it a try!

[tox]
minversion = 2.0
envlist = py{27,36,35,34}-sphinx{1.4,1.3,1.2},coverage,docs

[tox:travis]
2.7 = py27, docs
3.6 = py36, coverage
3.5 = py35
3.4 = py34

[testenv]
basepython =
    py27:     python2.7
    py36:     python3.6
    py35:     python3.5
    py34:     python3.4
    coverage: python3.6
# force Python 2.7 for generate documentation -- Python 3 will fail for
# unknown reason -- see TODO "Known issues - testing". Error is:
# Traceback (most recent call last):
#   File ".../sphinx/cmdline.py", line 244, in main
#     app.build(opts.force_all, filenames)
#   File ".../sphinx/application.py", line 291, in build
#     self.builder.build_all()
#   File ".../sphinx/builders/__init__.py", line 211, in build_all
#     self.build(None, summary='all source files', method='all')
#   File ".../sphinx/builders/__init__.py", line 265, in build
#     self.doctreedir, self.app))
#   File ".../sphinx/environment.py", line 569, in update
#     self._read_serial(docnames, app)
#   File ".../sphinx/environment.py", line 589, in _read_serial
#     self.read_doc(docname, app)
#   File ".../sphinx/environment.py", line 812, in read_doc
#     pickle.dump(doctree, f, pickle.HIGHEST_PROTOCOL)
# TypeError: can't pickle generator objects
    docs:     python2.7
deps =
# force Sphinx 1.2.2 -- in 1.2.3 we would run into issue #1585:
# "Autosummary of modules broken in Sphinx-1.2.3" -- only fixed
# in 1.3b1 -- there is no 1.2.4 release, see:
# - https://github.com/sphinx-doc/sphinx/issues/1585
# - http://www.sphinx-doc.org/en/1.4.8/changes.html#release-1-3b1-released-oct-10-2014
#   sphinx1.2: sphinx >= 1.2.0, <= 1.2.9999
    sphinx1.2: sphinx >= 1.2.0, <= 1.2.2
# force sphinxcontrib-programoutput 0.8.0 in the case of Sphinx 1.2 since
# newer versions do not support Sphinx older than Sphinx 1.3.5 -- in 0.10.0
# we would run into issue #17 -- see:
# - https://github.com/NextThought/sphinxcontrib-programoutput/issues/17
# - https://github.com/NextThought/sphinxcontrib-programoutput/blob/0.9/CHANGES.rst
    sphinx1.2: sphinxcontrib-programoutput >= 0.8.0, <= 0.8.0
# force sphinxcontrib-tikz 0.4.2 in the case of Sphinx 1.2 since newer versions
# of Sphinx do not support the API 'add_latex_package' that was introduced with
# Sphinx 1.3 -- in 0.4.4 we would run into an attribute error issue "object has
# no attribute 'add_latex_package'" -- see:
# - https://bitbucket.org/philexander/tikz/commits/139d419
# - http://www.sphinx-doc.org/en/stable/extdev/appapi.html
#   --> Sphinx.add_latex_package
    sphinx1.2: sphinxcontrib-tikz >= 0.4.2, <= 0.4.2
# force at least to Sphinx 1.3.5 -- in 1.3.4 and lower we would
# run into the sphinxcontrib-programoutput issue #17 -- see above:
    sphinx1.3: sphinx >= 1.3.5, <= 1.3.9999
    sphinx1.4: sphinx >= 1.4.0, <= 1.4.9999
    coverage:  sphinx >= 1.4.0, <= 1.4.9999
    docs:      sphinx >= 1.4.0, <= 1.4.9999
install_command =
#   http://tox.testrun.org/en/latest/config.html#confval-install_command=ARGV
    pip install {opts} {packages} {toxinidir}[dev,test]
setenv =
    LANG=C
passenv=
#   http://tox.testrun.org/en/latest/config.html#confval-passenv=SPACE-SEPARATED-GLOBNAMES
    TRAVIS*
changedir =
    docs:                         {toxinidir}/docs
commands=
    sphinx{1.4,1.3,1.2}:          check-manifest {toxinidir}
    sphinx{1.4,1.3,1.2}:          flake8 {toxinidir}
# nosetests will fail inside the py{36,35,34}-sphinx1.2 environment
# for unknown reason -- exclude it until found a fix. Error is:
# Traceback (most recent call last):
#   File ".../sphinx_testing/util.py", line 139, in decorator
#     app = TestApp(*self.sphinxargs, **sphinxkwargs)
#   File ".../sphinx_testing/util.py", line 78, in __init__
#     warning, freshenv, warningiserror, tags)
#   File ".../sphinx/application.py", line 139, in __init__
#     self._init_builder(buildername)
#   File ".../sphinx/application.py", line 201, in _init_builder
#     self.emit('builder-inited')
#   File ".../sphinx/application.py", line 400, in emit
#     results.append(callback(self, *args))
#   File ".../sphinx/ext/autosummary/__init__.py", line 526, in process_generate_options
#     base_path=app.srcdir)
#   File ".../sphinx/ext/autosummary/generate.py", line 110, in generate_autosummary_docs
#     items = find_autosummary_in_files(sources)
#   File ".../sphinx/ext/autosummary/generate.py", line 227, in find_autosummary_in_files
#     lines = f.read().splitlines()
#   File "...ii.py", line 26, in decode
#     return codecs.ascii_decode(input, self.errors)[0]
# UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1338:
#                     ordinal not in range(128)
    py27:                         nosetests --where=tests/functional
    py{36,35,34}-sphinx{1.4,1.3}: nosetests --where=tests/functional
    coverage:                     nosetests --with-coverage --where=tests/unit
    coverage:                     python {toxinidir}/run_coveralls_on_travis.py
    docs:                         sphinx-build -b html -d {envtmpdir}/doctrees . {envtmpdir}/html
    docs:                         sphinx-build -b latex -d {envtmpdir}/doctrees . {envtmpdir}/latex
    docs:                         /usr/bin/make -C {envtmpdir}/latex PDFLATEX=xelatex
#   docs:                         /usr/bin/make BUILDDIR={envtmpdir} html latexpdf
